File: content/11-graphs/rotting_oranges.js
function orangesRotting(grid) {
    let ROWS = grid.length, COLS = grid[0].length;
    let queue = [];
    let fresh = 0, time = 0;

    for (let r = 0; r < ROWS; r++) {
        for (let c = 0; c < COLS; c++) {
            if (grid[r][c] === 1) fresh++;
            else if (grid[r][c] === 2) queue.push([r, c]);
        }
    }

    let directions = [[0,1],[0,-1],[1,0],[-1,0]];
    while (queue.length > 0 && fresh > 0) {
        let len = queue.length;
        for (let i = 0; i < len; i++) {
            let [r, c] = queue.shift();
            for (let [dr, dc] of directions) {
                let row = r + dr, col = c + dc;
                if (row >= 0 && row < ROWS && col >= 0 && col < COLS && grid[row][col] === 1) {
                    grid[row][col] = 2;
                    queue.push([row, col]);
                    fresh--;
                }
            }
        }
        time++;
    }

    return fresh === 0 ? time : -1;
}

const assert = require('assert');
const g1 = [[2,1,1],[1,1,0],[0,1,1]];
assert.strictEqual(orangesRotting(g1), 4);
console.log("All tests passed!");

File: content/11-graphs/rotting_oranges.py
from collections import deque

def orangesRotting(grid) -> int:
    ROWS, COLS = len(grid), len(grid[0])
    q = deque()
    fresh = 0
    time = 0

    for r in range(ROWS):
        for c in range(COLS):
            if grid[r][c] == 1:
                fresh += 1
            elif grid[r][c] == 2:
                q.append((r, c))

    directions = [[0,1],[0,-1],[1,0],[-1,0]]
    while q and fresh > 0:
        for _ in range(len(q)):
            r, c = q.popleft()
            for dr, dc in directions:
                row, col = r + dr, c + dc
                if 0 <= row < ROWS and 0 <= col < COLS and grid[row][col] == 1:
                    grid[row][col] = 2
                    q.append((row, col))
                    fresh -= 1
        time += 1

    return time if fresh == 0 else -1

if __name__ == "__main__":
    g1 = [[2,1,1],[1,1,0],[0,1,1]]
    assert orangesRotting(g1) == 4, "Test 1 Failed"
    print("All tests passed!")
